### PR TITLE
Add testing for methods defined in GitGud object

### LINKED ISSUE #197: Check error codes in test suites
When commands fail, they should fail with status code 1, which could help us make sure users are never seeing an error that's coming from python.

### PR DESCRIPTION
**Pull request by sahansk2**
Fixes #197 

@benthayer Could you verify for which methods you would like tests?

### CODE REVIEW

File: gitgud/test_user.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import pytest
+
+from gitgud.__main__ import GitGud
+
+def test_assert_initialized():
+    pass
+def test_handle_commit():
+    pass
+def test_handle_contrib():
```
Comment (benthayer) [inline gitgud/test_user.py:9]: No test

File: gitgud/test_user.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    pass
+def test_handle_debug():
```
Comment (benthayer) [inline gitgud/test_user.py:11]: No test

File: gitgud/test_user.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import pytest
+
+from gitgud.__main__ import GitGud
+
+def test_assert_initialized():
```
Comment (benthayer) [inline gitgud/test_user.py:5]: Test in git gud repo, git repo, empty directory

File: gitgud/test_user.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    pass
+def test_handle_skills():
+    pass
+def test_handle_solution():
+    pass
+def test_handle_status():
+    pass
+def test_handle_test():
+    pass
+def test_is_initialized():
```
Comment (benthayer) [inline gitgud/test_user.py:43]: Test with assert_initilized

File: gitgud/test_user.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import pytest
+
+from gitgud.__main__ import GitGud
+
+def test_assert_initialized():
+    pass
+def test_handle_commit():
```
Comment (benthayer) [inline gitgud/test_user.py:7]: Since we might be removing this, no test yet unless you want to

File: gitgud/test_user.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    pass
+def test_handle_contrib():
+    pass
+def test_handle_debug():
+    pass
+def test_handle_explain():
```
Comment (benthayer) [inline gitgud/test_user.py:13]: Ideally, we test explain/goal/status on all levels and at all stages during the solving, but to keep things lightweight, let's maybe just do it for one level as a sanity check?

We could run the commands on all levels, but that is more time-consuming. What are your thoughts?

File: gitgud/test_user.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    pass
+def test_handle_goal():
+    pass
+def test_handle_help():
+    pass
+def test_handle_init():
+    pass
+def test_handle_issues():
```
Comment (benthayer) [inline gitgud/test_user.py:21]: No test

File: gitgud/test_user.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    pass
+def test_handle_contrib():
+    pass
+def test_handle_debug():
+    pass
+def test_handle_explain():
+    pass
+def test_handle_goal():
+    pass
+def test_handle_help():
```
Comment (benthayer) [inline gitgud/test_user.py:17]: No test

File: gitgud/test_user.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    pass
+def test_handle_init():
```
Comment (benthayer) [inline gitgud/test_user.py:19]: Test in empty dir/ non-empty dir/ init-ed dir

File: gitgud/test_user.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    pass
+def test_handle_issues():
+    pass
+def test_handle_level():
+    pass
+def test_handle_levels():
```
Comment (benthayer) [inline gitgud/test_user.py:25]: level/levels/skill - test in initilized dir/ uninit-ed dir

File: gitgud/test_user.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    pass
+def test_handle_load():
```
Comment (benthayer) [inline gitgud/test_user.py:27]: test load next/prev and test loading some generic level (or levels if you want).

Also test outside of an init-ed dir to make sure it doesn't throw an error

File: gitgud/test_user.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    pass
+def test_handle_reset():
+    pass
+def test_handle_show():
+    pass
+def test_handle_show_tree():
```
Comment (benthayer) [inline gitgud/test_user.py:33]: Test output in git dir/non-git dir

File: gitgud/test_user.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    pass
+def test_handle_level():
+    pass
+def test_handle_levels():
+    pass
+def test_handle_load():
+    pass
+def test_handle_reset():
+    pass
+def test_handle_show():
```
Comment (benthayer) [inline gitgud/test_user.py:31]: Test with show-tree

File: gitgud/test_user.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    pass
+def test_handle_show_tree():
+    pass
+def test_handle_skills():
+    pass
+def test_handle_solution():
```
Comment (benthayer) [inline gitgud/test_user.py:37]: Same tests as other user facing commands, but obviously this one, you'll need to test the --confirm flag

File: gitgud/test_user.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+    pass
+def test_handle_status():
+    pass
+def test_handle_test():
```
Comment (benthayer) [inline gitgud/test_user.py:41]: Load up a level and make sure that it it doesn't fail either when the level has been completed or when it hasn't

Also make sure it doesn't fail if we're not in a git dir

File: gitgud/test_user.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+    pass
+def test_is_initialized():
+    pass
+def test_load_level():
+    pass
+def test_parse():
```
Comment (benthayer) [inline gitgud/test_user.py:47]: Test `git gud` without any subcommands in git dir and non-git dir

File: gitgud/test_user.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+    pass
+def test_handle_solution():
+    pass
+def test_handle_status():
+    pass
+def test_handle_test():
+    pass
+def test_is_initialized():
+    pass
+def test_load_level():
```
Comment (benthayer) [inline gitgud/test_user.py:45]: `commits.csv` should be modified after this func is run, and the level file should correspond to the one loaded.

There should be a test for this in `test_operator.py` if there isn't one already

**Review by benthayer — COMMENTED** [review]
I was a bit quick with the comments here to try and be expedient. I think everything is covered. Let me know if you need clarification.

Also, at this point, I think creating levels is the priority, not necessarily the tests, so don't feel like you're obligated to write all these tests

**Comment by benthayer** [comment]
It might (probably will) make sense to split this up into multiple PRs

**Comment by sahansk2** [comment]
If you'd like multiple PRs, would you also like multiple issues for this? E.g. "Test user methods using webbrowser", "Test user methods related to loading", "Test assert initialized methods", etc?

**Comment by benthayer** [comment]
Yes, I think that makes sense. Obviously, there would be many issues, and writing the issues might take as long as it takes to simply write the test, so do what you feel is right. There should either be a pull request or an issue for each test

**Comment by sahansk2** [comment]
What we could do is make a new "meta-issue", similar to what #238 is. Each checkbox could be a method in the GitGud object.

Going off of this, we could maintain another branch `user_tests` on this repository, and people open up testing PRs to merge to `user_tests`. Once all the tests are in and everything looks set, we merge `user_tests` into `master`. This is similar to what the `intro` branch (or the `add` branch, I forget) was like a while back.

I'm not sure what's best, so it's your call. As an aside, should this PR be closed?

**Comment by benthayer** [comment]
I think we can open a meta-issue (along with other issues)

I don't think we need to make a separate branch for the tests, I think we can merge them right into master when they are done. 

This PR can stay open until there exists other issues/PRs that make use of the useful comments that are here

**Comment by benthayer** [comment]
Stale/closing

**Pull request closed without merging**